ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,474 @@
+    fn as_int(&self) -> ArbBitInt {
+        assert!(self.all_known());
+        self.ones.clone()
+    }
+
+    fn all_known(&self) -> bool {
+        self.unknowns.count_ones() == 0
+    }
+
+    /// Return an integer where the known bits are set
+    fn knowns(&self) -> ArbBitInt {
+        self.unknowns.bitneg()
+    }
+
+    /// Returns an integer where the places that are known zeros have a bit set
+    fn zeroes(&self) -> ArbBitInt {
+        self.knowns().bitand(&self.ones.bitneg())
+    }
+
+    fn known_ones(&self) -> ArbBitInt {
+        self.knowns().bitand(&self.ones)
+    }
+
+    fn and(&self, other: &KnownBitValue) -> KnownBitValue {
+        assert_eq!(self.ones.bitw(), self.unknowns.bitw());
+        assert_eq!(self.ones.bitw(), other.ones.bitw());
+        assert_eq!(self.unknowns.bitw(), other.unknowns.bitw());
+        let set_ones = self.ones.bitand(&other.ones);
+        let set_zeroes = self.zeroes().bitor(&other.zeroes());
+        let unknowns = self
+            .unknowns
+            .bitor(&other.unknowns)
+            .bitand(&set_zeroes.bitneg());
+        KnownBitValue {
+            ones: set_ones,
+            unknowns,
+        }
+    }
+
+    fn or(&self, other: &KnownBitValue) -> KnownBitValue {
+        assert_eq!(self.ones.bitw(), self.unknowns.bitw());
+        assert_eq!(self.ones.bitw(), other.ones.bitw());
+        assert_eq!(self.unknowns.bitw(), other.unknowns.bitw());
+        let set_ones = self.ones.bitor(&other.ones);
+        let unknowns = self
+            .unknowns
+            .bitor(&other.unknowns)
+            .bitand(&set_ones.bitneg());
+        KnownBitValue {
+            ones: set_ones,
+            unknowns,
+        }
+    }
+}
+
+/// Known-bits analysis.
+pub(super) struct KnownBits {
+    iidx_to_known_bits: IndexVec<InstIdx, Option<KnownBitValue>>,
+    curr_value_pending_commit: Option<KnownBitValue>,
+}
+
+impl PassT for KnownBits {
+    fn feed(&mut self, opt: &mut PassOpt, inst: Inst) -> OptOutcome {
+        self.curr_value_pending_commit = None;
+        match inst {
+            Inst::And(x) => self.opt_and(opt, x),
+            Inst::Const(x) => self.opt_const(x),
+            Inst::Or(x) => self.opt_or(opt, x),
+            _ => OptOutcome::Rewritten(inst),
+        }
+    }
+    fn inst_committed(&mut self, _opt: &CommitInstOpt, _iidx: InstIdx, _inst: &Inst) {
+        self.iidx_to_known_bits
+            .push(self.curr_value_pending_commit.clone());
+        assert_eq!(_iidx.index(), self.iidx_to_known_bits.len() - 1);
+    }
+}
+
+impl KnownBits {
+    /// Create an empty known bits analysis object.
+    pub(super) fn new() -> Self {
+        KnownBits {
+            iidx_to_known_bits: IndexVec::new(),
+            curr_value_pending_commit: None,
+        }
+    }
+
+    /// Returns what we known about the bits of `iidx`.
+    fn as_knownbits(&self, opt: &PassOpt, iidx: InstIdx) -> KnownBitValue {
+        if let Some(bits) = self.iidx_to_known_bits[iidx].clone() {
```

Comment:
I think this is `self.iidx_to_known_bits[iidx].unwrap_or_else(|| ...)` in disguise? Does clippy spot that?

Reply:
No it doesn't. I changed it in b69981a3cf7816458ce29fe6336e6b5a2a439157